`_testing.assert_artists_equal()` does not compare artists properly

This has a similar root cause as in #2281 : the test uses zip to compare two artist arrays but in the case one array is empty and the other isn't, no assertion will fail although artists arrays differ. Identified this while working on #2328 but I didn't want to fix it there since it's out of scope for that PR. A quick fix involves asserting arrays length equality first. Trying to add this comparison results with a few failures:
```
FAILED seaborn/tests/test_axisgrid.py::TestJointPlot::test_kde - AssertionError
FAILED seaborn/tests/test_axisgrid.py::TestJointPlot::test_scatter_hue - AssertionError
FAILED seaborn/tests/test_axisgrid.py::TestJointPlot::test_palette - AssertionError
FAILED seaborn/tests/test_distributions.py::TestDisPlot::test_with_rug[kwargs0] - AssertionError
FAILED seaborn/tests/test_distributions.py::TestDisPlot::test_with_rug[kwargs2] - AssertionError
```

--- a/seaborn/_testing.py
+++ b/seaborn/_testing.py
@@ -40,6 +40,7 @@ BAR_PROPS = [
 
 def assert_artists_equal(list1, list2, properties):
 
+    assert len(list1) == len(list2)
     for a1, a2 in zip(list1, list2):
         prop1 = a1.properties()
         prop2 = a2.properties()

--- a/seaborn/tests/test_axisgrid.py
+++ b/seaborn/tests/test_axisgrid.py
@@ -1515,8 +1515,8 @@ class TestJointPlot:
 
         g2 = ag.JointGrid()
         scatterplot(data=long_df, x="x", y="y", hue="a", ax=g2.ax_joint)
-        histplot(data=long_df, x="x", hue="a", ax=g2.ax_marg_x)
-        histplot(data=long_df, y="y", hue="a", ax=g2.ax_marg_y)
+        kdeplot(data=long_df, x="x", hue="a", ax=g2.ax_marg_x, fill=True)
+        kdeplot(data=long_df, y="y", hue="a", ax=g2.ax_marg_y, fill=True)
 
         assert_plots_equal(g1.ax_joint, g2.ax_joint)
         assert_plots_equal(g1.ax_marg_x, g2.ax_marg_x, labels=False)
@@ -1571,9 +1571,9 @@ class TestJointPlot:
         g1 = ag.jointplot(data=long_df, x="x", y="y", kind="kde")
 
         g2 = ag.JointGrid()
-        histplot(data=long_df, x="x", y="y", ax=g2.ax_joint)
-        histplot(data=long_df, x="x", ax=g2.ax_marg_x)
-        histplot(data=long_df, y="y", ax=g2.ax_marg_y)
+        kdeplot(data=long_df, x="x", y="y", ax=g2.ax_joint)
+        kdeplot(data=long_df, x="x", ax=g2.ax_marg_x)
+        kdeplot(data=long_df, y="y", ax=g2.ax_marg_y)
 
         assert_plots_equal(g1.ax_joint, g2.ax_joint)
         assert_plots_equal(g1.ax_marg_x, g2.ax_marg_x, labels=False)
@@ -1610,9 +1610,9 @@ class TestJointPlot:
         g1 = ag.jointplot(x="x", y="y", **kws)
 
         g2 = ag.JointGrid()
-        kdeplot(x="x", y="y", ax=g2.ax_joint, **kws)
-        kdeplot(x="x", ax=g2.ax_marg_x, **kws)
-        kdeplot(y="y", ax=g2.ax_marg_y, **kws)
+        scatterplot(x="x", y="y", ax=g2.ax_joint, **kws)
+        kdeplot(x="x", ax=g2.ax_marg_x, fill=True, **kws)
+        kdeplot(y="y", ax=g2.ax_marg_y, fill=True, **kws)
 
         assert_plots_equal(g1.ax_joint, g2.ax_joint)
         assert_plots_equal(g1.ax_marg_x, g2.ax_marg_x, labels=False)

--- a/seaborn/tests/test_distributions.py
+++ b/seaborn/tests/test_distributions.py
@@ -2039,15 +2039,13 @@ class TestDisPlot:
 
         ax = rugplot(data=long_df, **kwargs)
         g = displot(long_df, rug=True, **kwargs)
-        for bar in g.ax.patches:
-            bar.remove()
+        g.ax.patches = []
 
         assert_plots_equal(ax, g.ax, labels=False)
 
         long_df["_"] = "_"
         g2 = displot(long_df, col="_", rug=True, **kwargs)
-        for bar in g2.ax.patches:
-            bar.remove()
+        g2.ax.patches = []
 
         assert_plots_equal(ax, g2.ax, labels=False)
 

